telegram/ext/dispatcher.py

```diff
@@ -247,78 +247,51 @@ def _pooled(self):
                     promise.pooled_function.__name__)
                 continue
 
+            # We don't like infinite loops
             if promise.pooled_function in self.error_handlers:
-                self.logger.error('An error was raised while processing the update '
-                                  'and an uncaught error was raised while handling '
-                                  'the error with an error_handler')
+                self.logger.error('An uncaught error was raised while handling the error.')
                 continue
 
-            if promise.error_handler:
-                try:
-                    context = CallbackContext.from_error(promise.update, promise.exception,
-                                                         self)
-                    if promise.update:
-                        promise.error_handler(promise.update, context, *promise.args,
-                                              **promise.kwargs)
-                    else:
-                        promise.error_handler(context, *promise.args, **promise.kwargs)
-                except Exception:
-                    self.logger.exception('An error was raised while processing the update '
-                                          'and an uncaught error was raised while handling '
-                                          'the error with an error_handler')
-            elif promise.update:
-                try:
-                    self.dispatch_error(promise.update, promise.exception)
-                except Exception:
-                    self.logger.exception('An error was raised while processing the update '
-                                          'and an uncaught error was raised while handling '
-                                          'the error with an error_handler')
-            else:
-                self.logger.exception(
-                    'No error handlers are set for async function, logging exception.',
-                    exc_info=promise.exception)
+            # We still don't like infinite loops, but here a different error message is needed
+            if not promise.error_handling:
+                self.logger.error('A promise with deactivated error handling raised an error.')
+                continue
 
-    def run_async(self, func, *args, update=None, error_handler=None, **kwargs):
-        """Queue a function (with given args/kwargs) to be run asynchronously.
+            # If we arrive here, an exception happened in the promise and was neither
+            # DispatcherHandlerStop nor raised by an error handler. So we can and must handle it
+            try:
+                self.dispatch_error(promise.update, promise.exception, promise=promise)
+            except Exception:
+                self.logger.exception('An uncaught error was raised while handling the error.')
+
+    def run_async(self, func, *args, update=None, error_handling=True, **kwargs):
```

Comment:
I think that `error_handling` is an internal implementation detail and needs to be hidden from the user.
When registering error handlers we need to proactive set `error_handling` to `False` instead of leaving it to the hands of the user.

Reply:
I disagree on this one. Wen can't prevent users from calling `run_async` from within error handlers and at some point, someone will do so leading to an infinite loop. So we either document what `error_handling` is good for or *at least* make a note that using `run_async` from within error handlers is forbidden.